kubeadm/v1.15.0/KubeClusterHelper.psm1

```diff
@@ -0,0 +1,1353 @@
+        if ($LastExitCode) {
+            throw "Failed to pull $Global:PauseImage"
+        }
+    }
+}
+
+function InstallKubernetesBinaries()
+{
+    Param(
+    [parameter(Mandatory = $true)] $Source,
+    $DestinationPath
+    ) 
+
+    $existingPath = [Environment]::GetEnvironmentVariable("Path", [System.EnvironmentVariableTarget]::Machine)
+
+    # For current shell Path update
+    $env:path += ";$DestinationPath\kubernetes\node\bin"
+    # For Persistent across reboot
+    [Environment]::SetEnvironmentVariable("Path", $existingPath + ";$DestinationPath\kubernetes\node\bin", [EnvironmentVariableTarget]::Machine)
+
+    $env:KUBECONFIG = $(GetKubeConfig)
+    [Environment]::SetEnvironmentVariable("KUBECONFIG", $(GetKubeConfig), [EnvironmentVariableTarget]::Machine)
+
+    $Release = "1.15"
+    if ($Source.Release)
+    {
+        $Release = $Source.Release
+    }
+    $Url = "https://dl.k8s.io/v${Release}/kubernetes-node-windows-amd64.tar.gz"
+    if ($Source.Url)
+    {
+        $Url = $Source.Url
+    }
+    DownloadFile -Url  $Url -Destination $Global:BaseDir/kubernetes-node-windows-amd64.tar.gz
+    tar -zxvf $Global:BaseDir/kubernetes-node-windows-amd64.tar.gz -C $DestinationPath
+}
+
+function UninstallKubernetesBinaries()
+{
+    Param(
+    $DestinationPath
+    ) 
+    Remove-Item Env:\KUBECONFIG -ErrorAction SilentlyContinue
+
+    # For current shell Path update
+    $existingPath = [Environment]::GetEnvironmentVariable("Path", [System.EnvironmentVariableTarget]::Machine)
+    $existingPath = $existingPath.Replace($DestinationPath+'\bin;', "")
+    # For Persistent across reboot
+    [Environment]::SetEnvironmentVariable("Path", $existingPath, [EnvironmentVariableTarget]::Machine)
+    Remove-Item $DestinationPath -Force -ErrorAction SilentlyContinue
+}
+
+function InstallContainersRole()
+{
+    $feature = Get-WindowsFeature -Name Containers
+    if (!$feature.Installed)
+    {
+        Install-WindowsFeature -Name Containers -IncludeAllSubFeature
+        return $true
+    }
+
+    return $false
+}
+
+function ReadKubeClusterInfo()
+{
+    $KubeConfiguration = @{
+        ClusterCIDR = GetClusterCidr;
+        ServiceCIDR = GetServiceCidr;
+        KubeDnsIp = GetKubeDnsServiceIp;
+        NetworkName = $Global:NetworkName;
+        NetworkMode = $Global:NetworkMode;
+    }
+
+    $Global:Configuration += @{
+        Kube = $KubeConfiguration;
+    }
+    WriteKubeClusterConfig
+}
+
+function GetKubeDnsServiceIp()
+{
+    $svc = ConvertFrom-Json $(kubectl.exe get services -n kube-system -o json | Out-String)
+    $svc.Items | foreach { $i = $_; if ($i.Metadata.Name -match "dns") { return $i.spec.ClusterIP } }
+}
+
+function GetAPIServerEndpoint() {
+    $endpoints = ConvertFrom-Json $(kubectl.exe get endpoints --all-namespaces -o json | Out-String)
+    $endpoints.Items | foreach { $i = $_; if ($i.Metadata.Name -match "kubernetes") { return "$($i.subsets.addresses.ip):$($i.subsets.ports.port)" } }
```

Comment:
I had trouble using this with kubernetes-dashboard being deployed in my cluster. Below was how I fixed it locally.

```suggestion
    $endpoints.Items | foreach { $i = $_; if ($i.Metadata.Name -eq "kubernetes") { return "$($i.subsets.addresses.ip):$($i.subsets.ports.port)" } }
```